hypershift-operator/controllers/nodepool/manifests.go

```diff
@@ -15,27 +15,27 @@ import (
 	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-func machineDeployment(nodePool *hyperv1.NodePool, clusterName string) *capiv1.MachineDeployment {
+func machineDeployment(nodePool *hyperv1.NodePool, clusterName string, controlPlaneNamespace string) *capiv1.MachineDeployment {
```

Comment:
The namespace of all these resources strictly derives from a NodePool by design. Adding the ability for a consumer to break that seems incorrect to me.

Reply:
The control plane namespace is built and created [entirely as a function of only the HostedCluster](https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go#L219), not the NodePool. So anything in the system that wants to put stuff into the control plane namespace must get a reference to the namespace via the HostedCluster.

And a NodePool only exists in the context of a HostedCluster — if you have a NodePool, you necessarily have the HostedCluster.

Therefore it seems that any resource in the control plane namespace derived from the NodePool still must necessarily use the HostedCluster to calculate at least one property of the resource: its namespace (the control plane namespace, derived from HostedCluster).

Reply:
>Therefore it seems that any resource in the control plane namespace derived from the NodePool still must necessarily use the HostedCluster to calculate at least one property of the resource: its namespace (the control plane namespace, derived from HostedCluster).

This was programatically enforced before by invoking
```
func targetNamespace(nodePool *hyperv1.NodePool) string {
	return fmt.Sprintf("%s-%s", nodePool.GetNamespace(), nodePool.Spec.ClusterName)
}
```
within the func scope of the NodePool derived resources, whereas now is not, it's only suggested with parameter name.

No need to bikeshed on this though I'm good either way.

Reply:
What I mean is, the `targetNamespace` function in your example is now a second source of truth for the definition of the control plane namespace. A HostedCluster exists independent of any NodePool, but a NodePool can only exist in the context of a HostedCluster. You can have a control plane with no NodePools. The ControlPlaneNamespace function is the source of truth to get at the control plane namespace, and it requires the HostedCluster. The fact the name is derived from the HostedCluster name is an internal detail. Reproducing that naming function in `targetNamespace` here leaks the internal detail and establishes the NodePool as a second source of truth for the control plane namespace.

So my change intends to conform to a single source of truth principle for the control plane namespace.

I do agree there's something to improve here though in terms of the ergonomics/legibility... some other ideas:

* Make these functions accept both the NodePool and the HostedCluster (and internalize the ControlPlaneNamespace calls to get the namespace based on the HostedCluster).
* Remove namespace handling entirely from these functions and instead set them from the call side which could then set the namespace based on the ControlPlaneNamespace.

... or maybe something else?

Like you said, maybe better to just take a look at this particular issue separately later on.